ui/seek_bar.js

```diff
@@ -605,6 +607,225 @@ shaka.ui.SeekBar = class extends shaka.ui.RangeElement {
+    }
+
+    if (chapters.length < 2) {
+      return;
+    }
+
+    // Create chapter elements
+    const chaptersContainer = shaka.util.Dom.createHTMLElement('div');
+    chaptersContainer.classList.add('shaka-chapters');
+
+    const totalSize = chapters.reduce((t, c) => {
+      t += c.size;
+      return t;
+    }, 0);
+
+    /**
+     * @type {!Array<{
+     *  start: number,
+     *  end: number,
+     *  el: HTMLElement
+     * }>}
+     * */
+    const chapterElMap = [];
+
+    for (const c of chapters) {
+      /** @type {!HTMLElement} */
+      const chapterEl = shaka.util.Dom.createHTMLElement('div');
+      chapterEl.classList.add('shaka-chapter');
+      chapterEl.style.width = `${c.size * 100 / totalSize}%`;
+
+      chaptersContainer.appendChild(chapterEl);
+
+      /** @type {!HTMLElement} */
+      const chapterMarker = shaka.util.Dom.createHTMLElement('div');
+      chapterMarker.style.borderColor =
+        this.config_.seekBarColors.chapterMarks;
+      chapterEl.appendChild(chapterMarker);
+
+      /** @type {!HTMLElement} */
+      const chapterLabel = shaka.util.Dom.createHTMLElement('p');
+      chapterLabel.classList.add('shaka-chapter-label', hiddenClass);
+      chapterLabel.style.color =
+        this.config_.seekBarColors.chapterLabels;
+      chapterLabel.innerText = c.title;
+      chapterEl.appendChild(chapterLabel);
+
+      chapterElMap.push(
+          {start: c.start, end: c.end, el: chapterLabel});
+    }
+
+    // Remove chapter elements on abort
+    this.eventManager.listenOnce(abortSignal, 'abort',
+        () => {
+          this.recursivelyRemove_(chaptersContainer);
+        }, {passive: true});
+
+    // Add chapter event listeners
+    /** @type {!AddEventListenerOptions} */
+    const chapterEventOptions = {passive: true, signal: abortSignal};
+
+    this.eventManager.listen(this.bar, 'pointermove', (e) => {
+      if (!e.target) {
+        return;
+      }
+      const target = /** @type {HTMLElement} */(e.target);
+
+      const screenXDiff = e.offsetX / target.clientWidth;
+      const rangeMax = parseInt(target.getAttribute('max'), 10);
+      const hoverVal = screenXDiff * rangeMax;
+
+      for (const c of chapterElMap) {
+        const hidden = c.el.classList.contains(hiddenClass);
+        const inChapter = c.start <= hoverVal && hoverVal < c.end;
+        if (inChapter === hidden) {
+          c.el.classList.toggle(hiddenClass);
+        }
+      }
+    }, chapterEventOptions);
+
+    this.eventManager.listen(this.bar, 'pointerout', () => {
+      for (const c of chapterElMap) {
+        if (!c.el.classList.contains(hiddenClass)) {
+          c.el.classList.add(hiddenClass);
+        }
+      }
+    }, chapterEventOptions);
+
+    container.insertBefore(
+        chaptersContainer, this.container.childNodes[0]);
+  }
+
+  /**
+   * @param {Element} parent
+   * @private
+   */
+  recursivelyRemove_(parent) {
+    if (!parent) {
+      return;
+    }
+    while (parent.lastChild) {
```

Comment:
Is there actually a practical reason why we need to recursively remove all of the contents of the container? Would it not be good enough to just remove the chapter container?

Reply:
I have rewritten this part to make it simpler.